include/zfs_comutil.h

```diff
@@ -41,6 +41,27 @@ extern int zfs_spa_version_map(int zpl_version);
 #define	ZFS_NUM_LEGACY_HISTORY_EVENTS 41
 extern const char *zfs_history_event_names[ZFS_NUM_LEGACY_HISTORY_EVENTS];
 
+/*
+ * Truncating snprintf() without compiler warnings
+ *
+ * GCC 7.1 with "-Wall" will warn in cases where we're not checking the
+ * snprintf() return code when the buffer could be truncated:
+ *
+ *     libzfs_iter.c: In function ‘zfs_iter_bookmarks’:
+ *     libzfs_iter.c:207:40: error: ‘snprintf’ output may be truncated
+ *     before the last format character [-Werror=format-truncation=]
+ *         (void) snprintf(name, sizeof (name), "%s#%s", zhp->zfs_name,
+ *
+ * In many cases we don't care and just want to silently truncate the buffer.
+ *
+ * This hack fools the compiler by making the size volatile, so that it isn't
+ * able to compute the truncation at compile time, and thus no error.
+ */
+#define	SNPRINTF_TRUNC(str, size, ...) do {				\
+	volatile int __s = size;					\
+	snprintf(str, __s, ##__VA_ARGS__);				\
+} while (0)
```

Comment:
And with the only semi-legitimate usage of this now occurring in `ztest.c` we should be able to eliminate the need for this macro entirely by building `ztest` with `-Wnoformat-truncation`.  Here's an untested patch, you may need to structure this like the `ZFS_AC_CONFIG_USER_FRAME_LARGER_THAN` to ensure the version of gcc being used supports the option.  Alternately a [pragma](https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html) could be used for this.

```diff
diff --git a/cmd/ztest/Makefile.am b/cmd/ztest/Makefile.am
index a2f3b5a..c7ac681 100644
--- a/cmd/ztest/Makefile.am
+++ b/cmd/ztest/Makefile.am
@@ -1,6 +1,6 @@
 include $(top_srcdir)/config/Rules.am
 
-AM_CFLAGS += $(DEBUG_STACKFLAGS) $(FRAME_LARGER_THAN)
+AM_CFLAGS += $(DEBUG_STACKFLAGS) $(FRAME_LARGER_THAN) -Wnoformat-truncation
 AM_CPPFLAGS += -DDEBUG
 
 DEFAULT_INCLUDES += \
```